katdal/lazy_indexer.py

```diff
@@ -32,61 +32,59 @@
+    if not increment or increments_left:
+        raise ValueError('Could not convert {} to a slice (non-uniform or '
+                         'zero increments)'.format(index))
+    return slice(index[0], index[-1] + increment, increment)
 
-    In the current implementation, boolean numpy arrays which select contiguous
-    ranges are converted to slices. Note that when indexing along multiple axes
-    with arrays, this may change the semantics of the indexing (see NumPy's
-    `NEP 21`_ for details).
 
-    .. _NEP 21: http://www.numpy.org/neps/nep-0021-advanced-indexing.html
+def _simplify_index(indices, shape):
+    """Generate an equivalent index expression that is cheaper to evaluate.
 
-    If the expression is invalid in some way (e.g. wrong length array index)
-    the original index is returned so that that an exception will be raised
-    when the index is used.
+    Advanced ("fancy") indexing using arrays/lists of booleans or ints is much
+    slower than basic indexing using slices and scalar ints in dask. If the
+    fancy index on a specific axis/dimension selects a range with a fixed
+    (non-zero) step size between indices, however, it can be converted into an
+    equivalent slice to get a simple index instead.
 
-    Ellipses are currently not supported, and will result in the original
-    index being returned.
+    Note that when indexing along multiple axes with arrays, this may change
+    the semantics of the indexing (see NumPy's `NEP 21`_ for details). This
+    simplification is only guaranteed to be safe when used with outer indexing.
+
+    .. _NEP 21: http://www.numpy.org/neps/nep-0021-advanced-indexing.html
     """
-    if not isinstance(indices, tuple):
-        indices = (indices,)
+    # First clean up and check indices, unpacking ellipsis and boolean arrays
+    indices = da.slicing.normalize_index(indices, shape)
     out = []
     axis = 0
     for index in indices:
-        if index is Ellipsis:
-            return indices     # Not supported
-        elif index is np.newaxis:
-            out.append(index)
-        else:
-            if axis >= len(shape):
-                return indices   # Invalid index expression
+        if index is not np.newaxis:
             length = shape[axis]
             axis += 1
-            try:
-                bool_array = (index.dtype == np.bool_ and index.shape == (length,))
-            except AttributeError:
-                bool_array = False
-            if bool_array:
-                selected = np.nonzero(index)[0]
-                if not len(selected):
-                    index = slice(0, 0)
-                elif selected[-1] - selected[0] + 1 == len(selected):
-                    index = slice(selected[0], selected[-1] + 1)
-            out.append(index)
+            # If there is 1-D fancy index on this axis, try to convert to slice
+            if isinstance(index, np.ndarray) and index.ndim == 1:
```

Comment:
What would happen if someone passed a non-integer, non-boolean ndarray in? While it won't be legal, we ideally want to have it fail in the same way after `_simplify_index`.

Reply:
Both NumPy and dask's checks result in an `IndexError`. I've added a test for this, showing that `_simplify_index` and NumPy have matching errors.